internal/lsp/server.go

```diff
@@ -0,0 +1,522 @@
+package lsp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/sourcegraph/jsonrpc2"
+	"gopkg.in/yaml.v3"
+
+	"github.com/styrainc/regal/pkg/config"
+)
+
+type LanguageServerOptions struct {
+	ErrorLog       *os.File
+	VerboseLogging bool
+}
+
+func NewLanguageServer(opts *LanguageServerOptions) *LanguageServer {
+	ls := &LanguageServer{
+		cache:                      NewCache(),
+		errorLog:                   opts.ErrorLog,
+		diagnosticRequestFile:      make(chan fileDiagnosticRequiredEvent, 10),
+		diagnosticRequestWorkspace: make(chan string, 10),
+		verboseLogging:             opts.VerboseLogging,
+	}
+
+	return ls
+}
+
+type LanguageServer struct {
+	cache *Cache
+
+	conn *jsonrpc2.Conn
+
+	errorLog       *os.File
+	verboseLogging bool
+
+	loadedConfig     *config.Config
+	loadedConfigLock sync.Mutex
+
+	diagnosticRequestFile      chan fileDiagnosticRequiredEvent
+	diagnosticRequestWorkspace chan string
+
+	clientRootURI string
+}
+
+func (l *LanguageServer) Handle(ctx context.Context, conn *jsonrpc2.Conn, req *jsonrpc2.Request) (result interface{}, err error) {
+	l.logInboundMessage(req.Method, string(*req.Params))
+
+	if req.Params == nil {
+		return nil, &jsonrpc2.Error{Code: jsonrpc2.CodeInvalidParams}
+	}
+
+	switch req.Method {
+	case "initialize":
+		return l.handleInitialize(ctx, conn, req)
+	case "initialized":
+		return l.handleInitialized(ctx, conn, req)
+	case "textDocument/diagnostic":
+		return l.handleTextDocumentDiagnostic(ctx, conn, req)
+	case "textDocument/didOpen":
+		return l.handleTextDocumentDidOpen(ctx, conn, req)
+	case "textDocument/didClose":
+		return nil, nil
+	case "textDocument/didChange":
+		return l.handleTextDocumentDidChange(ctx, conn, req)
+	case "workspace/didChangeWatchedFiles":
+		return l.handleWorkspaceDidChangeWatchedFiles(ctx, conn, req)
+	case "workspace/diagnostic":
+		return l.handleWorkspaceDiagnostic(ctx, conn, req)
+	case "workspace/didRenameFiles":
+		return l.handleWorkspaceDidRenameFiles(ctx, conn, req)
+	case "workspace/didDeleteFiles":
+		return l.handleWorkspaceDidDeleteFiles(ctx, conn, req)
+	case "workspace/didCreateFiles":
+		return l.handleWorkspaceDidCreateFiles(ctx, conn, req)
+	}
+
+	return nil, &jsonrpc2.Error{Code: jsonrpc2.CodeMethodNotFound, Message: fmt.Sprintf("method not supported: %s", req.Method)}
+}
+
+func (l *LanguageServer) SetConn(conn *jsonrpc2.Conn) {
+	l.conn = conn
+}
+
+func (l *LanguageServer) StartDiagnosticsWorker(ctx context.Context) {
+	go func() {
```

Comment:
What's the benefit of having a go-func-only method here, as opposed to calling
```go
go l.StartDiagnosticsWorker(ctx)
```
wherever we call `l.StartDiagnosticsWorker(ctx)` now? 🤔